hibernate-core/src/main/java/org/hibernate/bytecode/enhance/spi/interceptor/EnhancementHelper.java

```diff
@@ -51,15 +58,38 @@ public static boolean includeInBaseFetchGroup(
 		if ( value instanceof ToOne ) {
 			final ToOne toOne = (ToOne) value;
 			if ( toOne.isLazy() ) {
-				if ( toOne.isUnwrapProxy() ) {
-					if ( toOne instanceof OneToOne ) {
-						return false;
+				// include it in the base fetch group so long as the config allows
+				// using the FK to create an "enhancement proxy"
+
+				// however a few special cases to handle...
+
+				if ( ! toOne.isReferenceToPrimaryKey() ) {
+					// we do not have a reference to the associated primary-key
+					return false;
+				}
+
+				if ( inheritanceChecker.hasSubclasses( toOne.getReferencedEntityName() ) ) {
+					// the associated type has subclasses - we cannot use the enhanced proxy and will generate a HibernateProxy
+					if ( ! toOne.isUnwrapProxyImplicit() ) {
+						// NO_PROXY was explicitly requested
+						BytecodeLogger.LOGGER.debugf(
+								"`%s#%s` was mapped with LAZY with explicit NO_PROXY but the associated entity (`%s`) has subclasses",
+								bootMapping.getPersistentClass().getEntityName(),
+								bootMapping.getName(),
+								toOne.getReferencedEntityName()
+						);
```

Comment:
I know it's a little late to ask this, but should this be a warning, or at least something a little stronger than just DEBUG logged, since `@LazyToOne(NO_PROXY)` is not compatible with an association with an entity class that has subclasses?